3.12/vm.md

```diff
@@ -0,0 +1,279 @@
+# VM definition
+
+## Abstract
+
+The core of the VM consists of a bytecode array, a stack, an array of locals, an array of names, an array of constants, and several special registers.
+In addition, there's infrastructure to handle call/return, yield/resume/throw, and await.
+This description is geared towards the 3.12 interpreter (whose design is very much in flux).
+
+## Data
+
+Note that some of these are raw arrays, whereas others are Python objects (typically tuples).
+
+- Bytecode: an array of ``_Py_CODEUNIT`` (16-bit) words containing instructions and inline caches. Accessed via ``next_instr``.
+- Stack: an array of ``PyObject`` pointers used as temporary values.
+  The stack grows up. The top is ``stack_pointer``.
+- "Fast" locals (``frame->f_localsplus``): an array of ``PyObject`` pointers used to hold arguments, local variables, and "cells" used for nonlocal variables.
+  Not to be confused with ``LOCALS()`` described below; these "fast" locals are used in regular functions.
+- Names (``names``, caching ``frame->f_code->co_names``): a tuple of string objects used for accessing globals and attributes.
+- Constants (``consts``, caching ``frame->f_code->co_consts``): a tuple of ``PyObject`` pointers holding constants referenced by the ``LOAD_CONST`` instruction.
+
+## Special registers
+
+- Frame pointer (``frame``, type ``_PyInterpreterFrame *``):
+  The current stack frame.
+  Other values can be found through this, though some are cached in other special registers.
+  The previous frame (where ``return`` or ``yield`` should go) is ``frame->previous``.
+- Instruction pointer (``next_instr``, type ``_Py_CODEUNIT *``):
+  Address of the next instruction (or sometimes the current one or its inline cache).
+  This points somewhere into the bytecode array.
+- Stack pointer (``stack_pointer``, type ``PyObject **``):
+  Address of the next free stack entry.
+  A push operation is ``*stack_pointer++ = x``; a pop operation is ``x = *--stack_pointer``.
+  The top value on the stack is ``stack_pointer[-1]``.
+- Eval breaker (``eval_breaker``, type ``_Py_atomic_int *``):
```

Comment:
Unrelated, but, since Python is C11 for a while now, do we need a type for an atomic int? Can we just use `atomic_int` instead?

Reply:
I have no idea. Every time I look at that code I am so scared I stop looking.

- https://github.com/python/cpython/blob/main/Include/internal/pycore_atomic.h
- https://github.com/python/cpython/blob/main/Include/internal/pycore_atomic_funcs.h

I'm not sure what the difference between the two files is.

Reply:
I know why: MSVC.  We added support for C11 atomics last month.  It's still too shiny in that neck of the woods to make this change yet. Maybe in a few years. :)